[WIP] Docker containers for a DAO seednode

mrosseel opened the pull request:
This PR adds docker containers and infrastructure to run a dao seednode.

Since some interest have been expressed to move to a docker image for all seednodes, this setup might be a good starting point for those efforts.

Have a look at the Known Issues and TODO section of the readme.md for open issues.

mrosseel committed:
Created docker-compose files to start a seednode

This starts a seednode - compiled from scratch, collectd+nginx
and a bitcoin full node.

mrosseel committed:
first checkin of new dao seednode docker files

mrosseel committed:
Adding rpcHost option, seednode docker fixes

mrosseel committed:
Merge branch 'rc_v1.0.0' of https://github.com/ManfredKarrer/bisq into dao-seednode

mrosseel committed:
checkout correct branch, fix rpchost option

mrosseel committed:
Add logs to assess correct functioning of the seed node

mrosseel committed:
fix dockerfile

mrosseel committed:
add host

mrosseel committed:
Merge branch 'rc_v1.0.0' of https://github.com/ManfredKarrer/bisq into dao-seednode

mrosseel committed:
limit btc node usage

mrosseel committed:
Clean up logging

mrosseel committed:
speed it up

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
protect against crashes

mrosseel committed:
Fix APP_NAME, revert logging/unnecessary edits

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
Storing backup files etc in local dir, not inside the container

mrosseel committed:
no more overlapping volumes

mrosseel committed:
fix notify bug

mrosseel committed:
btcd gracefully shuts down on certain errors so doesn't restart

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
add depends_on so notifications don't error out at startup

mrosseel committed:
Wait for bitcoin rpc port to be available before connecting

mrosseel committed:
correct path

mrosseel committed:
make executable

mrosseel committed:
increase timeout (just in case)

mrosseel committed:
switch dependency

mrosseel committed:
add while loop

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
fix quotes

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
remove stray line

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

# Conflicts:
#	core/src/main/java/bisq/core/dao/node/full/RpcService.java

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
completing readme for docker installation

mrosseel committed:
move to bitcoin 0.18, readme changes

mrosseel committed:
adding collectd, removing networking

ripcurlx commented:
@blabno @wiz @sqrrm Could someone with docker experience review this PR?

blabno commented on seednode/README.md:
```diff
@@ -17,3 +19,47 @@ Follow the logs created by the service by inspecting
+The production docker image uses the build dir, so make sure you have built the appropriate tag/branch.
+Here's the checklist to start your docker seednode:
+
+- build the seednode code: `./gradlew build`
```
This step should be part of image build process and build directory should be excluded.

mrosseel replied:
that would indeed be better, it used to be like that but the repos and branches kept changing :)

I could pass in the repo+branch via the .env file and checkout/build using that info. Of course that implies checking it out every time. Other suggestions welcome.

blabno commented:
Could this pr be squashed into single commit?

mrosseel commented:
> Could this pr be squashed into single commit?

sure, I'll squash it after processing the remarks and possibly making correction commits. Or do you prefer it squashed before?

blabno commented:
It's ok to do it in the end of the process.

freimair reviewed (changes requested):
I really wanna see the collectd reporting stuff integrated...

mrosseel commented:
> I really wanna see the collectd reporting stuff integrated...

that's what I'm working on now. It's code complete but not working yet. 
Wanted to keep it separate from this PR, collectd is after all a fairly recent requirement.
Let me know if you (and others) still want to see it in this PR, then we'll have to put this one on hold till the collectd stuff is done.

wiz commented:
I'm looking forward to testing this after it's all ready :)

mrosseel committed:
Working setup for collectd

Network, disk space are working, but reported to be smaller than
the actual host free disk space.

mrosseel committed:
Closed ports to docker only, moved certs to root of seednode, add btcd

mrosseel committed:
Correct the ip address of btcd

mrosseel committed:
Use builder docker, pass url/branch to indicate code to build

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into seednode-collectd

mrosseel committed:
Merge branch 'seednode-collectd' into dao-seednode

mrosseel committed:
Used gradle:jdk10 docker image to build the code

mrosseel committed:
Adapted rpcallowip to bitcoin core 0.18.1, removed wait-for-it.sh

mrosseel committed:
Take the start script from parent dir, not the one checked out

mrosseel committed:
Fixed copy of start script

mrosseel committed:
Added block notify host

mrosseel committed:
Modified rpc host

mrosseel committed:
Corrected blocknotify host

mrosseel committed:
Made startup work without copying everything, docker-compose cleanup

Made rpcport configurable, used container names where possible.

mrosseel committed:
Create own image for nginx

mrosseel committed:
Removed RPCService changes

mrosseel committed:
Add alpine-needed user

mrosseel committed:
Fix [alert] 8#8: connect() failed (2: No such file or directory)

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
Fixed conflict by using git pull instead of fetch

mrosseel committed:
Save commit from builder in a  file

mrosseel committed:
Merge remote-tracking branch 'upstream/master' into dao-seednode

mrosseel committed:
Merge branch 'dao-seednode' of github.com:mrosseel/bisq into dao-seednode

mrosseel committed:
Increase memory to stay in line with recommendations

mrosseel committed:
Increased Max memory to be in sync with previous memory commit

mrosseel committed:
Merge remote-tracking branch 'upstream/master' into dao-seednode

ripcurlx commented:
@mrosseel Why is this PR [WIP] are you waiting for specific feedback from someone?

mrosseel committed:
Updated readme for seednode

mrosseel committed:
Cleaned up startup script

mrosseel committed:
Merge branch 'master' of github.com:bisq-network/bisq into dao-seednode

mrosseel committed:
Upgrade gradle, gradle available for jdk11 is too old.

Should not be a problem because the jdk running the jar is still jdk10

mrosseel commented:
> @mrosseel Why is this PR [WIP] are you waiting for specific feedback from someone?

the collectd stuff is not working in prod (worked in testing) and would like @wiz to take a look. To be fair I've been waiting to ask him till I had time to follow up, which should be next week.

wiz commented on seednode/docker/prod/collectd.conf:
```diff
@@ -0,0 +1,145 @@
+Hostname "3ck4nimvkavyrgd5.onion"
```
```suggestion
```

wiz commented on seednode/docker/prod/nginx.conf:
```diff
@@ -0,0 +1,47 @@
+
+	server {
+ 		listen 2003;
+		allow 172.28.1.4;
+		deny all;
```
```suggestion
```

mrosseel committed:
Fixed bug with fixed onion address

wiz commented:
NACK - With all due respect for the hard work from @mrosseel that went into this PR, and after lots of consideration between the various strategies, my opinion is that Docker is not the best tool to install and maintain a Bisq seednode. 

I reviewed your Dockerfile setup and it looks okay, but I feel that Docker adds an unnecessary level of complexity, makes it difficult to customize the configuration of the daemons, and adds a security risk from the trusted nature of using third party docker images in the codebase.

I feel that it would be better for Bisq if we focus our efforts on the new [Bisq seednode installer](https://github.com/bisq-network/bisq/tree/master/seednode#bisq-seed-node), which allows a newbie to create a Bisq seednode with one single command, doesn't introduce any new trusted third parties, and is very transparent and easy to audit. Currently it's lacking `collectd` monitoring support, but that shouldn't be hard to add considering the entire installer was written in a day with help from @sqrrm.

I spent time reviewing your PR at your request - can I ask you to do the same and try out my new installer script and tell me what you think of it compared to your Docker approach?

pingiun commented on seednode/docker/startSeedNode.sh:
```diff
@@ -28,4 +54,14 @@ if [ ! -z "$USE_LOCALHOST_FOR_P2P" ]; then
+
+    echo `date` "(Re)-starting node"
+
+    JAVA_OPTS='-Xms8192M -Xmx8192M -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=6969 -Dcom.sun.management.jmxremote.rmi.port=6969 -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false' ./bin/bisq-seednode $ARGS 2>./error.log
```
I don't think jmx should be enabled by default

pingiun commented on seednode/docker/startSeedNode.sh:
```diff
@@ -28,4 +54,14 @@ if [ ! -z "$USE_LOCALHOST_FOR_P2P" ]; then
 fi
 
-JAVA_OPTS='-Xms1800m -Xmx1800m' ./build/app/bin/bisq-seednode $ARGS
+while true
```
Restarting shouldn't be builtin to the container, the docker itself can be restarted if needed

pingiun replied:
Instead of restarting for when bitcoind has not started yet, detect whether bitcoind has started and sleep when not

pingiun commented on seednode/docker/prod/openjdk.conf:
```diff
@@ -0,0 +1 @@
+/usr/lib/jvm/java-11-openjdk-amd64/lib/server
```
What is this used for?

pingiun commented on seednode/docker/prod/Dockerfile-btc:
```diff
@@ -0,0 +1,7 @@
+FROM ruimarinho/bitcoin-core:0.18
```
I think we should prefer building the bitcoin node in the bisq source tree. It's pretty easy to make a Dockerfile for bitcoind, so this dependency is unnecessary

pingiun commented on seednode/docker/prod/Dockerfile:
```diff
@@ -1,16 +1,27 @@
+FROM gradle:jdk11 as builder
+ARG SEEDNODE_URL
+ARG SEEDNODE_BRANCH
+RUN git clone $SEEDNODE_URL
```
This is already in the same repository as the seednode code, we can use local data instead of fetching it from github

pingiun commented on seednode/.dockerignore:
```diff
@@ -5,7 +5,7 @@ docs/
 docker-compose.yml
 docker/development/
-docker/prod/
+#docker/prod/
```
I think this line and docker/development can be removed?

pingiun commented on seednode/docker/prod/collectd.conf:
```diff
@@ -0,0 +1,144 @@
+	</Plugin>
+</Plugin>
+
+#<Plugin load>
```
Don't comment out lines, remove them

pingiun commented on seednode/docker-compose.yml:
```diff
@@ -2,21 +2,112 @@ version: '2.1'
+            bisq_net:
+                ipv4_address: 172.28.1.2
+
+    nginx:
```
Why is the nginx container necessary?
